velox/dwio/parquet/reader/IAAPageReader.cpp

```diff
@@ -0,0 +1,510 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/dwio/parquet/reader/IAAPageReader.h"
+#include "velox/dwio/common/BufferUtil.h"
+
+namespace facebook::velox::parquet {
+
+using thrift::Encoding;
+using thrift::PageHeader;
+
+void IAAPageReader::preDecompressPage(
+    bool& need_pre_decompress,
+    int64_t numValues) {
+  if (codec_ != thrift::CompressionCodec::GZIP) {
+    need_pre_decompress = false;
+    return;
+  }
+  for (;;) {
+    auto dataStart = pageStart_;
+    if (chunkSize_ <= pageStart_) {
+      // This may happen if seeking to exactly end of row group.
+      numRepDefsInPage_ = 0;
+      numRowsInPage_ = 0;
+      break;
+    }
+    PageHeader pageHeader = readPageHeader();
+    pageStart_ = pageDataStart_ + pageHeader.compressed_page_size;
+    switch (pageHeader.type) {
+      case thrift::PageType::DATA_PAGE:
+        prefetchDataPageV1(pageHeader);
+        break;
+      case thrift::PageType::DATA_PAGE_V2:
+        prefetchDataPageV2(pageHeader);
+        break;
+      case thrift::PageType::DICTIONARY_PAGE:
+        prefetchDictionary(pageHeader);
+        continue;
+      default:
+        break; // ignore INDEX page type and any other custom extensions
+    }
+    break;
+  }
+  need_pre_decompress = isWinSizeFit;
+  rowGroupPageInfo_.numValues = numValues;
+  rowGroupPageInfo_.visitedRows = 0;
+}
+
+void IAAPageReader::prefetchNextPage() {
```

Comment:
I'm wondering if the function is designed to prefetch only the page immediately following the current one. Also, I'm curious about the behavior when seekToPage potentially skips the next page. In such a scenario, does the function also discard the predecompressed data for that skipped page?

Reply:
- The data page prefetch and pre-decompression happens when there is existing page in the current row group. And the function is called after processing the current page include reading and seting 'rowOfPage_' and 'numRowsInPage_' and initializes a decoder in function seekToPreDecompPage
- The function doesn't discard the predecompressed data for that skipped page, the calling trace is PageReader::skip->IAAPageReader::seekToPage->IAAPageReader::prefetchNextPage

Reply:
Got it. Thanks for your detailed explanation!